Foundation/NSXMLElement.swift

```diff
@@ -7,165 +7,271 @@
+        }
+    }
+
     /*!
         @method initWithXMLString:error:
         @abstract Returns an element created from a string. Parse errors are collected in <tt>error</tt>.
     */
     public init(XMLString string: String) throws { NSUnimplemented() }
-    
-    public convenience override init(kind: NSXMLNodeKind, options: Int) { NSUnimplemented() }
-    
+
+    public convenience override init(kind: NSXMLNodeKind, options: Int) {
+        self.init(name: "", URI: nil)
+    }
+
     /*!
         @method elementsForName:
         @abstract Returns all of the child elements that match this name.
     */
-    public func elementsForName(name: String) -> [NSXMLElement] { NSUnimplemented() }
-    
+    public func elementsForName(name: String) -> [NSXMLElement] {
+        return self.filter({ _CFXMLNodeGetType($0._xmlNode) == _kCFXMLTypeElement }).filter({ $0.name == name }).flatMap({ $0 as? NSXMLElement })
+    }
+
     /*!
         @method elementsForLocalName:URI
         @abstract Returns all of the child elements that match this localname URI pair.
     */
     public func elementsForLocalName(localName: String, URI: String?) -> [NSXMLElement] { NSUnimplemented() }
-    
+
     /*!
         @method addAttribute:
         @abstract Adds an attribute. Attributes with duplicate names are not added.
     */
-    public func addAttribute(attribute: NSXMLNode) { NSUnimplemented() } //primitive
-    
+    public func addAttribute(attribute: NSXMLNode) {
+        guard _CFXMLNodeHasProp(_xmlNode, UnsafePointer<UInt8>(_CFXMLNodeGetName(attribute._xmlNode))) == nil else { return }
+        addChild(attribute)
+    } //primitive
+
     /*!
         @method removeAttributeForName:
         @abstract Removes an attribute based on its name.
     */
-    public func removeAttributeForName(name: String) { NSUnimplemented() } //primitive
-    
+    public func removeAttributeForName(name: String) {
+        let prop = _CFXMLNodeHasProp(_xmlNode, name)
+        if prop != nil {
+            let propNode = NSXMLNode._objectNodeForNode(_CFXMLNodePtr(prop))
+            _childNodes.remove(propNode)
+            // We can't use `xmlRemoveProp` because someone else may still have a reference to this attribute
+            _CFXMLUnlinkNode(_CFXMLNodePtr(prop))
+        }
+    } //primitive
+
     /*!
         @method setAttributes
         @abstract Set the attributes. In the case of duplicate names, the first attribute with the name is used.
     */
-    public var attributes: [NSXMLNode]?
-    
+    public var attributes: [NSXMLNode]? {
+        get {
+            var result: [NSXMLNode] = []
+            var attribute = _CFXMLNodeProperties(_xmlNode)
+            while attribute != nil {
+                result.append(NSXMLNode._objectNodeForNode(attribute))
+                attribute = _CFXMLNodeGetNextSibling(attribute)
+            }
+            return result.count > 0 ? result : nil // This appears to be how Darwin does it
```

Comment:
effectively that is the case; the attributes are stored as an ivar which is lazily constructed